### PR TITLE
Simplify config parsing

### PR DESCRIPTION
**Pull request by saschagrunert**
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:
We can assume that the config is a flat map of keys and values and manually merge them into the `Config` struct. This allows to avoid manually parsing the YAML.

#### Which issue(s) this PR fixes:

None
#### Special notes for your reviewer:
None
#### Does this PR introduce a user-facing change?

```release-note
None
```

### CODE REVIEW

File: pkg/common/file.go [hunk 1]
```diff
@@ -34,175 +35,123 @@ type Config struct {
...
-			if err != nil {
-				return nil, fmt.Errorf("parsing config option '%s': %w", name, err)
-			}
-		case "disable-pull-on-run":
-			config.DisablePullOnRun, err = strconv.ParseBool(value)
-			if err != nil {
-				return nil, fmt.Errorf("parsing config option '%s': %w", name, err)
-			}
-		default:
-			return nil, fmt.Errorf("Config option '%s' is not valid", name)
```
Comment (SergeyKanzhelev) [inline pkg/common/file.go:125]: so there will be no more check for unknown option? Cannot find it now

File: pkg/common/file.go [hunk 2]
```diff
@@ -34,175 +35,123 @@ type Config struct {
...
+	debugKey             = "debug"
+	pullImageOnCreateKey = "pull-image-on-create"
+	disablePullOnRunKey  = "disable-pull-on-run"
+)
+
 // ReadConfig reads from a file with the given name and returns a config or
 // an error if the file was unable to be parsed.
-func ReadConfig(filepath string) (*Config, error) {
-	data, err := os.ReadFile(filepath)
+func ReadConfig(path string) (*Config, error) {
```
Comment (SergeyKanzhelev) [inline pkg/common/file.go:51]: maybe easier to review if there are some unit tests for this function.

Comment (saschagrunert) [reply pkg/common/file.go:51]: I agree, we should add them before this PR.

/hold

Comment (SergeyKanzhelev) [reply pkg/common/file.go:51]: Also those edge cases: https://github.com/kubernetes-sigs/cri-tools/pull/1646/files#r1811995991

I am curious if parsing to map would have troubles with the duplicate entries and whether it will keep which value will be used.

Comment (saschagrunert) [reply pkg/common/file.go:52]: Rebased the PR and fixed the logic accordingly to the failing test.s

File: pkg/common/file.go [hunk 3]
```diff
@@ -34,177 +36,147 @@ type Config struct {
...
-	if len(yamlData.Content) == 0 ||
-		yamlData.Content[0].Content == nil {
-		return config, nil
-	}
-	contentLen := len(yamlData.Content[0].Content)
-
-	// YAML representation contains 2 yaml ScalarNodes per config option.
-	// One is config option name and other is the value of the option
-	// These ScalarNodes help preserve comments associated with
-	// the YAML entry
```
Comment (SergeyKanzhelev) [inline pkg/common/file.go:95]: I don't understand this comment. Why and where the comments associated with the YAML entry was ever used?

File: pkg/common/file.go (hunk 2 above)
Comment (SergeyKanzhelev) [reply pkg/common/file.go:52]: thank you!

File: pkg/common/file.go (hunk 3 above)
Comment (SergeyKanzhelev) [reply pkg/common/file.go:95]: besides this comment it is lgtm

Comment (saschagrunert) [reply pkg/common/file.go:95]: Looks like we cannot preserve the comments without the current logic. Closing this PR for now.

Comment (SergeyKanzhelev) [reply pkg/common/file.go:95]: I am not sure where we are preserving comments. Aren't config is just readonly?

If we do preserver comments in some scenario - it will be great to have a test for it so it is not just a comment.

Comment (saschagrunert) [reply pkg/common/file.go:95]: We preserve the comments but not the newlines right now, I added an e2e test scenario for the whole `crictl config` subcommand in https://github.com/kubernetes-sigs/cri-tools/pull/1659
